consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,488 @@
+know which participant a message came from and that messages cannot be forged.
+
+Only messages from participants are considered.
+
+For messages containing a value only messages with valid values are considered.
+
+Participants are able to broadcast messages to all other participants. In the
+case that a participant is off-line or somehow inaccessible they will not
+receive broadcast messages and there is no mechanism for these messages to be
+re-sent.
+
+We refer to a consensus instance to mean consensus for a specific height,
+consensus instances are independent (they have no shared state), that is
+messages or values from the previous height have no bearing on the outcome of
+the current height
+
+We assume the existence of an application that feeds values to be agreed upon
+to the consensus instance and also receives agreed upon values from the
+consensus. The application also provides application specific implementations
+for certain functions that the consensus algorithm relies upon, such as a
+function that determines the validity of a value. 
+
+### Algorithm
+
+See supporting [functions](#Functions) and [notation](#Appendix-1-Notation).
+
+Note all state modifications are contained in the following code block,
+supporting functions cannot modify state, they all operate without side
+effects.
+
+Variable names                  |Instance state                          
+--------------------------------|----------------------------------------
+`H - height`                    |`Hc - current height`                                      
+`R - round          `           |`Rc - current round`                    
+`V - value`                     |`Rd - desired round`                    
+`T - Message type`              |`Vc - currently proposed value`         
+`RCC - round change certificate`|`PCc - current prepared certificate`    
+`PC - prepared certificate`     |`Sc - current participant state`        
+`S - participant state`         |
+`M or N - message sets`         |
+
+```
+
+// Upon receiving this event the algorithm transitions to the next height the
+// final committed event signifies that the application has accepted the agreed
+// upon value for the current height.
+upon: <FinalCommittedEvent>
+  Hc ← Hc+1
+  Rc ← 0
+  Rd ← 0
+  Sc ← AcceptRequest
+  Vc ← nil
+  schedule onRoundChangeTimeout(Hc, 0) after roundChangeTimeout(0)
+
+// A request event is a request to reach agreement on the provided value, the
+// request event is sent by the application, if this parcicipant is the proposer
+// it will propose that value by sending a preprepared message.
+upon: <RequestEvent, Hc, V> && Sc = AcceptRequest
+  if Rc = 0 && isProposer(Hc, Rd) {
+    bc(<Preprepare, Hc, 0, V, nil>)
+  }
+
+// When receiving a preprepare from a proposer participants will vote for the
+// value (if valid) by sending a prepare message.
+upon: <Preprepare, Hc, Rd, V, RCC> from proposer(Hc, Rd) && Sc = AcceptRequest
```

Comment:
I found notation a bit confusing here:
Is this pattern matching: `<Preprepare, Hc, Rd, V, RCC>`, meaning that the height in the msg must be equal to `Hc` and the round in the msg should be equal `Rd`?

It would seem like that's the interpretation, but i'm not so sure

Reply:
You could call it pattern matching, the approach is described here - https://github.com/celo-org/celo-blockchain/blob/be8e779fb879e63cfae8f0d60f9b4f9ff3163fd3/consensus/istanbul/celo-ibft-spec.md?plain=1#L323-L325

Let me know how you would change it if that is not clear.